src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,333 @@
+            if (memory.IsEmpty)
+                return memory;
+
+            int ichLim = memory.Length;
+            var span = memory.Span;
+            if (!char.IsWhiteSpace(span[ichLim - 1]))
+                return memory;
+
+            while (0 < ichLim && char.IsWhiteSpace(span[ichLim - 1]))
+                ichLim--;
+
+            return memory.Slice(0, ichLim);
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(ReadOnlyMemory<char> memory, out Single value)
+        {
+            var res = DoubleParser.Parse(out value, memory.Span);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(ReadOnlyMemory<char> memory, out Double value)
+        {
+            var res = DoubleParser.Parse(out value, memory.Span);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        public static uint Hash(ReadOnlySpan<char> span, uint seed) => Hashing.MurmurHash(seed, span);
+
+        public static NormStr AddToPool(ReadOnlyMemory<char> memory, NormStr.Pool pool)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Add(memory);
+        }
+
+        public static NormStr FindInPool(ReadOnlyMemory<char> memory, NormStr.Pool pool)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Get(memory);
+        }
+
+        public static void AddToStringBuilder(ReadOnlyMemory<char> memory, StringBuilder sb)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+            if (!memory.IsEmpty)
+                sb.AppendAll(memory);
+        }
+
+        public static void AddLowerCaseToStringBuilder(ReadOnlyMemory<char> memory, StringBuilder sb)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+
+            if (!memory.IsEmpty)
+            {
+                int min = 0;
+                int j;
+                var span = memory.Span;
+                for (j = min; j < memory.Length; j++)
+                {
+                    char ch = CharUtils.ToLowerInvariant(span[j]);
+                    if (ch != span[j])
+                    {
+                        sb.Append(memory, min, j - min).Append(ch);
+                        min = j + 1;
+                    }
+                }
+
+                Contracts.Assert(j == memory.Length);
+                if (min != j)
+                    sb.Append(memory, min, j - min);
+            }
+        }
+
+        private static bool ContainsChar(char ch, char[] rgch)
+        {
+            Contracts.AssertNonEmpty(rgch, nameof(rgch));
+
+            for (int i = 0; i < rgch.Length; i++)
```

Comment:
This should be changed to `return rgch.AsSpan().IndexOf(ch) >= 0`. In the future it can just be `return rgch.AsSpan().Contains(ch)` after we get a build with [https://github.com/dotnet/coreclr/pull/19863](https://github.com/dotnet/coreclr/pull/19863) (i.e. .NET Core 3.0).

Reply:
Although, if you fix the caller to no longer call this method, it can just be deleted all together.

---
In reply to: [217071879](https://github.com/dotnet/machinelearning/pull/863#discussion_r217071879) [](ancestors = 217071879)